server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java

```diff
@@ -573,6 +587,9 @@ public void writeTo(StreamOutput out) throws IOException {
             } else {
                 out.writeBoolean(SnapshotsService.useShardGenerations(entry.version));
             }
+            if (out.getVersion().onOrAfter(DATA_STREAMS_IN_SNAPSHOT)) {
```

Comment:
We should assert that we don't have any data streams if we're writing to an older version and must make sure to not start snapshots that include data streams as long as there's old version nodes in the cluster that don't understand data stream snapshots.

Reply:
Added required check